src/theme/input-field.css.ts

```diff
@@ -39,3 +47,188 @@ export const inputFieldStyle = style({
+  {
+    display: "none",
+  },
+)
+
+globalStyle(
+  `${inputFieldVisibilityToggleStyle}:not(:checked) ~ label svg:first-child`,
+  {
+    display: "none",
+  },
+)
+
+/**
+ * Workaround for the custom input field with the security icon
+ * We want to pass along the state of the input field to the container
+ *
+ * Below we ensure the browser supports the :has selector, so this won't be applied to browsers that don't support it
+ */
+globalStyle(
+  `${passwordInputContainerStyle}:has(${inputFieldSecurityStyle}:not(:focus):not(:placeholder-shown):valid)`,
+  {
+    border: `1px solid ${oryTheme.success.emphasis}`,
+  },
+)
+
+/**
+ * Workaround for the custom input field with the security icon
+ * We want to pass along the state of the input field to the container
+ *
+ * Below we ensure the browser supports the :has selector, so this won't be applied to browsers that don't support it
+ */
+globalStyle(
+  `${passwordInputContainerStyle}:has(${inputFieldSecurityStyle}:not(:focus):not(:placeholder-shown):invalid)`,
+  {
+    border: `1px solid ${oryTheme.error.emphasis}`,
+  },
+)
+
+/**
+ * Workaround for the custom input field with the security icon
+ * We want to pass along the state of the input field to the container
+ *
+ * Below we ensure the browser supports the :has selector, so this won't be applied to browsers that don't support it
+ */
+globalStyle(
+  `${passwordInputContainerStyle}:has(:not(:focus):not(:placeholder-shown):valid)`,
+  {
+    border: `1px solid ${oryTheme.input.disabled}`,
+  },
+)
+
+/**
+ * On selecting the visibility checkbox, we want to show the text in the input field
+ *
+ **/
+globalStyle(
+  `${inputFieldVisibilityToggleStyle}:checked ~ ${inputFieldSecurityStyle}`,
+  {
+    ["-webkit-text-security"]: "none",
+  } as GlobalStyleRule & {
+    ["-webkit-text-security"]: "none"
+  },
+)
+
+/**
+ * Firefox does not support the :has selector yet, so we need to test for this
+ * https://caniuse.com/css-has
+ * https://caniuse.com/mdn-css_properties_-webkit-text-security
+ * If both the :has selector and the -webkit-text-security: disc are supported,
+ * we can show the password input field with a visibility toggle
+ * and hide the fallback password input field
+ **/
+globalStyle(`${passwordInputContainerStyle}`, {
+  "@supports": {
+    "(-webkit-text-security: disc) and selector(:has(a,b))": {
+      display: "flex",
+    },
+  },
+})
+
+/**
+ * Firefox does not support the :has selector yet, so we need to test for this
+ * https://caniuse.com/css-has
+ * https://caniuse.com/mdn-css_properties_-webkit-text-security
+ * If both the :has selector and the -webkit-text-security: disc are supported,
+ * we hide this input field in favor of the custom password input field
+ **/
+globalStyle(`${inputFieldStyle}[type="password"]`, {
+  "@supports": {
+    "(-webkit-text-security: disc) and selector(:has(a,b))": {
+      display: "none",
+    },
+  },
+})
```

Comment:
Not sure how to test this properly, but to my understanding, selectors that can't be understood are just not matched. Therefore, we don't need a `@supports` clause for `:has` and can instead just use the selector, so long as the fallback is well-setup and secure. Perhaps something like this?

```suggestion
/**
 * Firefox does not support the :has selector yet, so we need to test for this
 * https://caniuse.com/css-has
 * https://caniuse.com/mdn-css_properties_-webkit-text-security
 * If both the :has selector and the -webkit-text-security: disc are supported,
 * we can show the password input field with a visibility toggle
 * and hide the fallback password input field
 **/
globalStyle(`${passwordInputContainerStyle}:has(*)`, {
  "@supports": {
    "(-webkit-text-security: disc)": {
      display: "flex",
    },
  },
})

/**
 * Firefox does not support the :has selector yet, so we need to test for this
 * https://caniuse.com/css-has
 * https://caniuse.com/mdn-css_properties_-webkit-text-security
 * If both the :has selector and the -webkit-text-security: disc are supported,
 * we hide this input field in favor of the custom password input field
 **/
globalStyle(`${inputFieldStyle}[type="password"]:has(*)`, {
  "@supports": {
    "(-webkit-text-security: disc)": {
      display: "none",
    },
  },
})
```